Replace reliance on CONNECTIVITY_ACTION broadcasts

lognaturel opened the pull request:
Closes #2499

This PR narrowly converts the auto-send functionality to use `WorkManager` as suggested by @jd-alexander and started in #2511. Unlike #2511, it does not convert other jobs to `WorkManager`. 

This also increases the target API to 26 to close #1992 because the move away from using `CONNECTIVITY_ACTION` was the last change needed for that.

**Please review commit by commit**. The commits are organized to minimize review time. Most or all could be squashed for merge.

**Questions for reviewers**
- There's one fairly significant issue I know of: if there's an unhandled crash while the `AutoSendWorker` is running, the app needs to be uninstalled and reinstalled for autosend to work again, presumably because work stays enqueued but has no way of resuming. I have not looked at the intended way for `Worker`s to handle that yet.
- There is still a lot of possible cleanup here. Is there anything you see as necessary for a merge?

#### What has been done to verify that this works as intended?
All verifications have been done on Moto G5, Android 7.0
- [x] Manually submit to GSheets and server with valid settings
- [x] Finalize form with wifi on with both GSheets and server, verify auto-send happens with valid settings
- [x] Finalize form with wifi off with both GSheets and server, verify auto-send happens when wifi comes back
- [x] Set server polling interval to 15 minutes, turn wifi off, turn wifi on 17 minutes later and make sure server is polled (NOTE: this may not work for Android <= 4.4 but I would say that's ok)
- [x] Trigger auto-send with non-OpenRosa server (e.g. google.com) and verify error message is helpful
- [x] Trigger auto-send with HTTP server that can do HTTPS redirect and verify HTTPS is included in uriremap map
- [x] Verify upload buttons don't trigger upload if autosend is ongoing
- [ ] Send > 1000 submissions at once manually and auto for server and GSheets

#### Why is this the best possible solution? Were any other approaches considered?
See some of the collapsed notes below and @grzesiek2010's comment [here](https://github.com/opendatakit/collect/pull/2619#issuecomment-428938167) for why using `WorkManager` is probably our best bet. This PR extracts out the submission code for reuse between the `Worker` and `AsyncTask`s to make sure everything runs on expected threads. An ideal solution might remove the `AsyncTask`s entirely but that is even higher risk and we're out of time for the Play Store deadline.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Intentional changes:
- Google Sheets uploads can't be canceled part way through 
- Error messages for auth problems and Google Sheets configuration problems are more detailed
- URL errors and auth errors do not stop the entire submission because we don't know if they apply to all the forms left to send

Risks:
- Lots of code movement for both GSheets and server upload triggered manually or automatically
- Possible unintentional change in how submission error states are handled
- Database state not being updated correctly on sends
- Server polling not retrying on network availability on Android <= 4.4
- Problems related to targeting API 26. We don't know of any but will be particularly important to try things out on Android 8.0 because I don't have such a device.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
Possibly if server polling is not retried on Android <= 4.4

#### Before submitting this PR, please make sure you have:
- [ ] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### Overview of the approach:
- `FormEntryActivity.requestAutoSend` is called when a filled form is finalized and the form should be auto-sent. It enqueues a job with constraints that make sure it runs only when connectivity is available and that a maximum of one such job is enqueued at any time.

- The job is defined in `AutoSendWorker.doWork`.

- `InstanceServerUploader` and `InstanceGoogleSheetsUploader` are now simple objects that both the upload `AsyncTask`s and the `AutoSendWorker` rely on.

- `InstanceServerUploader` and `InstanceGoogleSheetsUploader` use `Instance` objects rather than passing around cursors and doing direct database calls.

- Both uploaders have a `uploadOneSubmission` method that uploads one submission and is the primary entry point. They provide results back to the callers by throwing `UploadException`s in case of failures.

<details><summary>Oct. 19 todos</summary>
- I was hoping to use the `InstanceUploader` class more but I'm not seeing a great way to make the `uploadOneSubmission` signature consistent between uploader types. Are there other good strategies I'm missing for promoting reuse (e.g. maintaining more state in the objects instead of passing things into the method)?

Remaining to do:
- [x] Block manual upload for server and GSheets if auto-send is ongoing
- [x] Use `UploadException` for server submission to match Google Sheets submission
- [x] Trigger form update check when network becomes available if last check failed (can probably do with constraint on `ServerPollingJob`)
- [ ] Reduce redundancy in `AutoSendWorker`. Possibly split into subclasses?
- [x] Provide better feedback for failure states not linked to instances. This is already a problem in `master` so not urgent.
- [x] Separate analytics for server auto-send
</details>

<details><summary>Oct. 8 original writeup</summary>
The goal is to evaluate the approach and make a decision on whether it's the way to go. To do this, I highlight a general overview of the approach and some concerns below.

I started by clarifying the intent behind some existing methods and breaking others down where feasible. I used Android Studio's refactor tools as much as possible. `NetworkReceiver` is no longer used but I left it in so all the refactors there can be viewed at once.

The conversion from `NetworkReceiver` to `AutoSendWorker` is nearly identical to what @jd-alexander did in #2511 with a few fixes and some inline `TODO`s with considerations.

Here's an overview of the approach:
- `FormEntryActivity.enqueueAutosendJobIfNeeded` is called when a filled form is finalized. If that filled form should be autosent, an autosend job is enqueued. Constraints on it make sure it runs only when connectivity is available and that a maximum of one such job is enqueued at any time.

- The job is defined in `AutoSendWorker.doWork`. It uses the existing instance uploader `AsyncTask`s to perform submissions.

- The Worker blocks while an `AsyncTask` performs its background work. A `CountdownLatch` is used to synchronize that.

- `AutoSendWorker.uploadingComplete` is called when the async tasks complete and decrements the latch to resume the Worker's work.

- The Worker result comes from a field that several different methods can set. The only case where there's a retry is if the connection type that just became available doesn't match the settings. Otherwise everything is success or failure.

Some concerns:
- This structure results in `AsyncTask`'s `execute` being called from a background thread [which is not allowed](https://developer.android.com/reference/android/os/AsyncTask#threading-rules). It's not clear what the implications are and I think this was already happening with the `NetworkReceiver` implementation. I believe the same issue would come up using `AlarmManager` and `JobScheduler` directly. The structure of the uploaders probably needs to be rethought.

- It's hard to track whether the countdown latch is always decremented where it should be and there's a risk of missing one and having work that never completes. Addressing the previous point could help with that if we can avoid calling an `AsyncTask`.

- There are cases where it might make sense to retry but there isn't enough information in the worker to decide that. For example, `uploadingComplete` gets localized messages for each filled form that was attempted to send. Those can't be used to see if a server is inaccessible and may be accessible later. Again, I think this is an existing problem and not inherent to this approach.

#### Questions for reviewers:
- Do you see a straightforward way to address any of the concerns above or reasons why they shouldn't be concerns?
- Are you inclined to continue in the direction of using `WorkManager` or have another suggestion?

#### Remaining to do:
- Block manual uploads while automatic upload is ongoing. There's an approach [here](https://github.com/opendatakit/collect/pull/2511/files#diff-a9c23be08ab0d8fb651e22afa05e8add).
- Poll server for form update immediately after connection is received. #2511 did this by converting the server polling job to use `WorkManager`. It's possible `android-job` also makes it possible to retry when connectivity is available, this should be verified.
</details>

lognaturel committed:
Introduce WorkManager dependency

Guava needs to be excluded because google-api-client already imports it.

lognaturel committed:
Put entry to task at top

lognaturel committed:
Rename values to instancesToUpload to clarify semantics

lognaturel committed:
Convert NetworkReceiver to WorkManager Worker

lognaturel committed:
Remove NetworkReceiver triggers from manifest

zestyping commented on collect_app/src/main/java/org/odk/collect/android/receivers/NetworkReceiver.java:
```diff
@@ -73,13 +74,23 @@ public void onReceive(Context context, Intent intent) {
+     * @return true if a connection is available and settings specify it should trigger auto-send,
+     * false otherwise.
+     */
+    private boolean networkTypeMatchesAutoSendSetting(NetworkInfo currentNetworkInfo) {
```
If I'm understanding this right, it seems a name like `isAutosendEnabledForNetworkType` would get across the idea better for me.  But let me know if I've misunderstood the intent.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/receivers/NetworkReceiver.java:
```diff
@@ -235,26 +250,35 @@ public void uploadingComplete(HashMap<String, String> result) {
+     * TODO: Verify that this works with localization and that there really are no other messages
+     * that can indicate success (e.g. a custom server message).
+     */
+    private boolean allFormsDownloadedSuccessfully(Map<String, String> resultsMessagesByInstanceId) {
```
Does this really mean `allForms`**Uploaded**`Successfully`?

zestyping commented on collect_app/src/main/java/org/odk/collect/android/receivers/NetworkReceiver.java:
```diff
@@ -94,9 +105,13 @@ private boolean isFormAutoSendOptionEnabled(NetworkInfo currentNetworkInfo) {
+     * autosend at the form level. If the app-level autosend setting is disabled, send all filled
+     * forms that specify to send at the form level.
+     *
+     * @param isAutoSendAppSettingEnabled whether the auto-send option is enabled at the app level
```
This is a bit nitpicky, but I'm noticing that "autosend" is sometimes capitalized like it's one word (`autosend`, `KEY_AUTOSEND`) and sometimes capitalized like it's two words (`autoSend` or `AutoSend`).  Since in prose it's mostly written as one word ("autosend") and only occasionally hyphenated ("auto-send") and never as two words ("auto send"), I think it would be ideal to capitalize it always as one word.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/receivers/NetworkReceiver.java:
```diff
@@ -61,20 +61,16 @@ public void onReceive(Context context, Intent intent) {
         NetworkInfo currentNetworkInfo = manager.getActiveNetworkInfo();
 
-        if (action.equals(ConnectivityManager.CONNECTIVITY_ACTION)) {
+        if (action.equals(ConnectivityManager.CONNECTIVITY_ACTION)
```
Phew 😄  I read this method a bunch of times trying to figure out why it was acquiring the ConnectivityManager twice.  Reassuring to know that it was just unnecessary duplication.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/receivers/NetworkReceiver.java:
```diff
@@ -272,6 +255,27 @@ private String formatOverallResultMessage(HashMap<String, String> resultMessages
+
+        NotificationManager notificationManager = (NotificationManager) Collect.getInstance()
+                .getSystemService(Context.NOTIFICATION_SERVICE);
+        notificationManager.notify(1328974928, builder.build());
```
1328974928?  What's that?

lognaturel replied:
Clearly, there's more cleanup that can be done. It's an ID unique to this notification (and should be a named value somewhere).

zestyping commented on collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:
```diff
@@ -2837,8 +2842,28 @@ protected void onStop() {
+     * happen either because the blank form specifies that it should be autosent or because app-
+     * wide-autosend is enabled.
+     *
+     * TODO: if the user changes autosend settings, should an autosend job immediately be enqueued?
```
I was wondering this too.  I'd like to know more about the thinking behind why it's better to schedule a job at form finalization time than to check for forms to send at autosend time.

lognaturel replied:
I think we definitely want to schedule the job at form finalization time because that form may need to be sent and there may not be an active job. For example, a job could have successfully sent all the forms last time connectivity came on, now connectivity is still available and so we want that one form to be sent right away.

In other words, finalizing a form is a trigger for autosending.

But scheduling a job right when the user turns on autosend also seems like a good idea to me.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/workers/AutoSendWorker.java:
```diff
@@ -18,79 +18,133 @@
 
     private String resultMessage;
+    /**
+     * Causes the doWork method to wait until the instanceUploader AsyncTask has completed.
```
Why can't the doWork method do the work synchronously?

lognaturel replied:
It absolutely should -- that's what I'm getting at about the need to restructure the uploaders. The code for uploading in the auto-send/auto send/autoSend case is the same as the code for manually uploading which is done by an `AsyncTask` currently. So the quickest way to get this autosend approach working was just to call the functionality from the `AsyncTask`. As I noted in the PR description, it means `execute` is called from a background thread when it shouldn't. One option would be to pull the upload functionality out into a separate class that both the worker and the asynctask use. I feel like there's probably a better structure to be had, though.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/receivers/NetworkReceiver.java:
```diff
@@ -61,20 +61,16 @@ public void onReceive(Context context, Intent intent) {
-                uploadForms(context, networkTypeMatchesAutoSendSetting(currentNetworkInfo));
+                // If we just got connectivity, poll the server for form updates if we are due for it
+                if (action.equals(ConnectivityManager.CONNECTIVITY_ACTION)) {
+                    ServerPollingJob.pollServerIfNeeded();
```
Based on my quick reading of the `pollServerIfNeeded` method, maybe it would be better named `onNetworkConnected`?  It has more of the flavour of an event handler.

zestyping reviewed (commented):
I read through these changes and jotted down some of my thoughts as I was going along, to remind me of things I was confused about or things to discuss when we look at this tomorrow.

lognaturel commented:
@grzesiek2010 I see you're looking at this which is awesome! Would be great to get your feedback on the high-level questions first. There are certainly lots more naming and usage improvements that can be made but the most important thing is to make sure we're going in the right overall direction. Thanks!

lognaturel commented:
@zestyping and I did a very useful interactive review of this. The main conclusion is to move forward with `WorkManager` and address the remaining 3 concerns. @grzesiek2010, if you have an alternate preference, please share. 

I have moved these two concerns out of the PR description because we decided the tradeoffs are worth it:
- `WorkManager` is an alpha library. I asked for feedback on production use from Google [here](https://issuetracker.google.com/issues/116843000) and didn't get much of a response (expected). Blog posts and videos make it sound like other organizations are using it in production. As far as I know, the only alternative is to implement separate `JobScheduler` and `AlarmManager` implementations. This is what `WorkManager` and `android-job` do under the hood with the addition of `JobDispatcher` which should be more efficient for API 19 and 20 than `AlarmManager`. `android-job` is not much help because it [doesn't offer Android 4.x support](https://github.com/evernote/android-job/issues/429) for connectivity-triggered jobs so that would have to be custom anyway.

- I'm not sure how `WorkManager` sets up connectivity-triggered jobs with `AlarmManager`. @zestyping has pointed out this could be quite inefficient. Presumably we could find out from [the source code](https://android.googlesource.com/platform/frameworks/support/+/androidx-master-dev/work/workmanager) though I'm not finding out from a quick browse. `JobDispatcher` is not open source so it's a total black box.

The next set of tasks for this PR will be to address the other concerns. @zestyping mentioned it would be easier to review a diff without the non-controversial refactors (first 8 commits) so I pulled them out to #2632 for separate review.

Quick notes:
```
Main priority: pull code out of AsyncTask to share between AsyncTask and Worker
	InstanceServerUploader would be that new class; current InstanceServerUploader would get renamed to InstanceServerUploaderTask
		- one entrypoint method that returns an outcome

	Can intent extra url/username/password just live in the AsyncTask?
		- passed in via constructor, made final
		- passed on to InstanceServerUploader also through constructor, hopefully 

Nice to haves:
	Pull out condition of whether to enqueue job from enqueueJobIfNeeded
	Rethink Outcome
		- how is auth failure propagated?
		- can the per-form error messages be returned rather than mutate
		- use error keys rather than localized strings
	InstanceUploaderListener
		- comment that it's used to update UI about progress
		- maybe rename to UploadingProgressUI?
		- how is it related to publishProgress in AsyncTask?
	does delete instance stuff get invoked from background (InstanceUploader.onPostExecute)
	maybe address SQLite issue by not using placeholders
		- or new structure may address problem
	If something bad happens during send, onPostExecute is never run so we have no record
	Fail fast from irrecoverable errors
```

lognaturel committed:
Add database ID to Instance

lognaturel committed:
Get all instances from instance cursor

lognaturel committed:
Use Instance objects to simplify logic

Abstracts away the SQL chunking and provides a higher level entity to pass on

lognaturel committed:
Pass instance in to upload method

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:
```diff
@@ -2532,6 +2543,25 @@ public void next() {
+     * Requests that unsent finalized forms be auto-sent. If no network connection is available,
+     * the work will be performed when a connection becomes available.
+     *
+     * TODO: if the user changes auto-send settings, should an auto-send job immediately be enqueued?
```
I would say it should but it's not very important.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/workers/AutoSendWorker.java:
```diff
@@ -0,0 +1,372 @@
+     * Returns true if at least one form currently on the device specifies that all of its filled
+     * forms should auto-send no matter the connection type.
+     *
+     * TODO: figure out where this should live
```
Maybe just in FormsDao.

grzesiek2010 commented:
> Are you inclined to continue in the direction of using WorkManager or have another suggestion?

I made a research, analyzed possible solutions and in my opinion `WorkManager` is a right way!

> This structure results in AsyncTask's execute being called from a background thread which is not allowed. It's not clear what the implications are and I think this was already happening with the NetworkReceiver implementation. I believe the same issue would come up using AlarmManager and JobScheduler directly. The structure of the uploaders probably needs to be rethought.

`BroadcastReceiver.onReceive()` is always called within the main thread so using AsyncTasks was reasonable. In case of `doWork()` it's a separate thread so maybe it's better to fix that in this pr instead of in a separate one and test everything twice. 

> It's hard to track whether the countdown latch is always decremented where it should be and there's a risk of missing one and having work that never completes. Addressing the previous point could help with that if we can avoid calling an AsyncTask.

Then your second concern disappears.


>There are cases where it might make sense to retry but there isn't enough information in the worker to decide that. For example, uploadingComplete gets localized messages for each filled form that was attempted to send. Those can't be used to see if a server is inaccessible and may be accessible later. Again, I think this is an existing problem and not inherent to this approach.

We can think about improvements here but it's not something we need in this pr. I think you can just fille a new issue after merging this pr.

lognaturel commented:
Thanks for the feedback so far! I've started separating the Worker and AsyncTask at https://github.com/opendatakit/collect/compare/master...lognaturel:collect-2499-play. I have added a database ID field to the `Instance` class. This makes it possible to remove some of the database complexity. I've moved some of the common functionality to the dreadfully named `InstanceServerUploaderFriend`. Instead of mutating an `Outcome` object to get submission results, I've introduced a new `UploadResult` class. This allows us to return more information about each submission. Auto-send to a server works well. Remaining work is to do something similar for Google Sheets, incorporate auto-delete behavior and name things sensibly.

I'm working in a separate branch because I'm not sure of the structure I want to end up with. Any feedback is much appreciated. The commits should tell a pretty good story starting from "Put entry to task at top".

lognaturel committed:
Simplify getting url, username, password, delete after send overrides

lognaturel committed:
Add comments about intent extras used as overrides

lognaturel committed:
Move request authorization code to GoogleAccountsManager

lognaturel committed:
Increase target SDK version to 26

The last known issue to resolve is tasks triggered by network availability

lognaturel committed:
Clarify submission URL override behavior

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/dao/InstancesDao.java:
```diff
@@ -292,10 +292,14 @@ public void deleteInstancesFromIDs(List<String> ids) {
         List<Instance> instances = new ArrayList<>();
         if (cursor != null) {
             try {
+                cursor.moveToPosition(-1);
```
@grzesiek2010 Was there any reason that you didn't reset the cursor here?

grzesiek2010 replied:
No, I can't recall anything.

lognaturel committed:
Append deviceId in getUrlToSubmitTo method

lognaturel committed:
Pull out methods to save upload status to the database

lognaturel committed:
Clarify and simplify comments and white space

lognaturel committed:
Remove unused uploadedInstances field

lognaturel committed:
Add comment about instance deletion in onPostExecute

lognaturel committed:
Return status for only that submission when uploading

The Outcome object is only relevant for the AsyncTask and this makes it possible to remove it as a parameter for performing the upload.

lognaturel committed:
Rename upload-related AsyncTasks to make clear that they are tasks

lognaturel committed:
Move auto-send worker and upload results to upload package

lognaturel committed:
Move upload functionality to plain object

lognaturel committed:
Get Instance objects that need to be auto-sent

lognaturel committed:
Pull getURLToSubmitTo out for reuse

lognaturel committed:
Remove Worker dependence on AsyncTask for server upload

This also allows the notification title to be based on whether there were any failure results rather than trying to parse the result strings.
Auto-delete is not implemented. Google Sheets upload doesn't work yet.

lognaturel committed:
Add auto-delete after auto-send for server upload

lognaturel committed:
Add InstanceUploader class and pull out methods for GSheets

lognaturel committed:
Rename values to instanceIdsToUpload to match server task

lognaturel committed:
Make method to get unique form def by id and version public

lognaturel committed:
Abstract away the SQL queries for getting instances to send

This matches the structure of the ServerUploader. There are some redundant parameters that will be cleaned up later. Passing an instance down to the various methods makes it possible to remove the jrformID field.

lognaturel committed:
Move GSheets upload logic for reuse between AsyncTask and Worker

Removes possibility to cancel an upload between rows. Ideally, the upload of each form would be atomic so this is an improvement. It also better matches the server upload.

lognaturel committed:
Pull auth verification and submission folder creation out of task

lognaturel committed:
Add GoogleSheets auto-send

lognaturel commented:
I believe this is now fully functional. I have updated the PR description with new status information and questions for reviewers.

As I went through this, I noticed some unexpected behavior for GSheets submissions that exists in `master`:
- media filenames include the full path from the device (e.g. `/storage/emulated/0/odk/instances/odk-sheets-demo_2018-10-16_12-29-46/1539718210344.jpg`). Previously it just included the filename (e.g. `1539718210344.jpg`)
- media columns include links to the files (e.g. https://drive.google.com/open?id=1yEnky66isUsQ06HFVGUrqK9AM63yjcAg) rather than the filename (e.g. `1539718210344.jpg`)

I haven't had a chance to see when these issues came in. @grzesiek2010, could you please take a look and file issues as appropriate since you were most recently working with that code? 🙏

lognaturel committed:
Specify that an auth request gets all instances attempted

Ideally the auth request could be decoupled from submission attempt state but the database does not provide enough information to know which instances to filter out.

lognaturel committed:
Disable manual send when auto-send is ongoing

lognaturel committed:
Use android-job requirement to poll server on network availability

Allows removal of NetworkReceiver

lognaturel committed:
Split server analytics for auto and manual send

lognaturel committed:
Add copyright headers

lognaturel committed:
Throw exceptions on server upload failure to match GSheets

lognaturel committed:
Define common interface for InstanceUploaders

lognaturel committed:
Pull out submission code shared by uploader types

Also improve the way general (not instance-specific) failures are handled. This allows us to remove the resultMessage field which was just used for an auth failure that is no longer considered fatal (because the next form could not have it) and the corresponding string.

lognaturel committed:
Save the submission status to the db in the uploaders

So we can guarantee the status is always updated

lognaturel committed:
Narrow down the HEAD request exception catching

We don't want to catch the UploadExceptions we may have thrown. Also improves some of the error messages.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceServerUploader.java:
```diff
@@ -103,7 +103,17 @@ protected Outcome doInBackground(Long... instanceIdsToUpload) {
             String urlString = getURLToSubmitTo(instance, deviceId, completeDestinationUrl);
 
-            if (!uploadOneSubmission(instance, urlString, uriRemap, outcome)) {
+            SubmissionUploadResult result = uploadOneSubmission(instance, urlString, uriRemap);
```
Please note that this structure gets changed in 4d89018 to throw exceptions rather than return a result. This better matches the way GSheets uploading works.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/upload/AutoSendWorker.java:
```diff
@@ -301,13 +276,9 @@ private boolean atLeastOneFormSpecifiesAutoSend() {
-                    ? resultMessage
-                    : Collect.getInstance().getString(R.string.odk_auth_auth_fail);
-        } else {
+        if (resultMessagesByInstanceId != null) {
```
What made `R.string.odk_auth_auth_fail` go away?

lognaturel replied:
The `resultMessage` field used to exist for non-form-specific failures which were considered fatal. If that notification method was called with a null messages map, it was assumed that an auth error happened (which I don't think was a good assumption).

Now, auth errors are not considered fatal -- there's just a message shown for the form that it happened for. Any other irrecoverable non-form-specific failures like not having a Google user set when using the gsheets "protocol" results in a message being passed directly to the notification message instead of using a field.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceServerUploaderTask.java:
```diff
@@ -78,27 +79,23 @@ protected Outcome doInBackground(Long... instanceIdsToUpload) {
+            } catch (UploadAuthRequestedException e) {
+                outcome.authRequestingServer = e.getSubmissionUri();
+                // Don't add the instance that caused an auth request to the map because we want to
+                // retry; items in the map are not reconsidered in this submission attempt
```
Wording here is a little confusing; better would be e.g.

"items present in the map are considered already attempted and won't be retried"

zestyping commented on collect_app/src/main/java/org/odk/collect/android/upload/UploadAuthRequestedException.java:
```diff
@@ -0,0 +1,30 @@
+
+import android.net.Uri;
+
+public class UploadAuthRequestedException extends UploadException {
```
This class could use a docstring to explain what this exception means.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/upload/UploadException.java:
```diff
@@ -12,7 +12,7 @@
+package org.odk.collect.android.upload;
 
 /**
  * An exception that results in the cancellation of an instance upload, and the presentation of an
```
This is a good place to say a bit about how the various possibilities are represented: success, failure, fatal failure, retry (i.e. `UploadException` is anything non-successful, doesn't distinguish between failure and fatal failure, and there's only one kind of retryable failure, `UploadAuthRequestedException`).

zestyping commented on collect_app/src/main/java/org/odk/collect/android/upload/UploadException.java:
```diff
@@ -0,0 +1,33 @@
+ * An exception that results in the cancellation of an instance upload, and the presentation of an
+ * error to the user
+ * */
+public class UploadException extends Exception {
```
At the moment it looks like the result of `getDisplayMessage()` (if not also the `message` field itself) is a mixture of localizable and non-localizable strings, since sometimes a localized message is passed in and sometimes an exception.  Also, in the various places where an `UploadException` is constructed from another exception, sometimes it's done with `new UploadException(e)` and sometimes `new UploadException(e.getMessage())`.

I think it's okay to postpone addressing the localization issue for now, though I think it's important to document that the mixing of localizable and non-localizable strings is an intentional but temporary state of affairs.

(Presumably the future step would be to add a `new UploadException(int messageId)` constructor.)

I would also suggest always constructing with `new UploadException(e)` and not `new UploadException(e.getMessage())` so that we can preserve information we might use later to decide what to do about localization.

zestyping commented:
This looks pretty good to me; I'm happy to see some of the simplifications from the last round, and the whole thing is definitely a lot cleaner overall.

One more thing I'd suggest is doc strings on the fields of `Outcome`, since there are some detailed semantics packed in there.  Particularly:

  - On the `messagesByInstanceId` field, the fact that the presence of a key-value pair indicates that an upload attempt was made in this particular "session" or "round" of uploading, and shouldn't be retried in the current session/round; also, the semantics of the messages (where do they come from, are they localizable/user-facing etc.)
  - What the `authRequestingServer` field means exactly, when it gets populated etc.

Beyond that and the other comments I've posted, I'm not seeing any significant concerns!

lognaturel commented:
Thanks a million, @zestyping! Will address your comments. I will also need to add a commit or two to get the tests running again. Seems that the version of Robolectric we had been using is not compatible with targeting API 26 and that when I upgrade it I run into some dependency incompatibility issues.

shobhitagarwal1612 commented on collect_app/src/main/java/org/odk/collect/android/upload/AutoSendWorker.java:
```diff
@@ -0,0 +1,322 @@
+        FormsDao dao = new FormsDao();
+        Cursor cursor = dao.getFormsCursor();
+
+        try {
```
`try finally` statement can be converted into a `try with resources` statement

lognaturel replied:
@grzesiek2010 did mention that elsewhere and although I would like to make that change, I can't get it to compile. https://developer.android.com/studio/write/java8-support#supported_features suggests that Android Studio 3.0 extends support to lower minSdks.

shobhitagarwal1612 replied:
Yes, I was able to make the change using Android Studio's auto fix

shobhitagarwal1612 commented on collect_app/src/main/java/org/odk/collect/android/upload/InstanceUploader.java:
```diff
@@ -0,0 +1,123 @@
+        Cursor cursor = new FormsDao().getFormsCursorForFormId(jrFormId);
+        String autoDelete = null;
+        if (cursor != null && cursor.moveToFirst()) {
+            try {
```
`try finally` can be converted into `try with resources`
If the initialized resource (here `cursor`) is null then the auto-closable isn't called ([link](https://stackoverflow.com/a/35372185))

lognaturel committed:
Upgrade robolectric to 4.0-beta-1, add Guava

Tests hang with both Robolectric 3.5.1 and 3.8. The explicit Guava dependency is because android-api-client imported a really old version that did not provide the methods Robolectric needs.

The Guava 27 changelog says 'Important: ListenableFuture is now available as a separate artifact (com.google.guava:listenablefuture:1.0) from the rest of Guava. As part of that change, Guava now has a real (needed at runtime) dependency on com.google.guava:failureaccess:1.0'. With 27, we were getting a conflict with failureaccess. Guava 27 is really new so better to just go to a version that works for now.

lognaturel committed:
Make location validating method static for easier testing

lognaturel committed:
Use try-with-resources

lognaturel committed:
Move Outcome to InstanceUploaderTask and add comments

The Outcome object is only used for upload tasks.

lognaturel committed:
Improve comments for UploadException

kkrawczyk123 commented:
There are some problems that I encountered:
- Notifications and action of periodic form update don’t work at all on 8.1.

- Only on one device Moto G Play, Android 6.0 I saw an issue connected to differentiate the type of connection. On cellular only in autosend - submission was sent when I turn wifi on not cellular and the same in the opposite situation when autosend was set to wifi only and I turn on cellular data. I wasn't able to reproduce it on other devices or on the same device but on the master branch. 

- I also have found some suspicious behavior: I did not add creds to aggregate for which it is required - auto-sending failed so I changed for Google service and try to send again failed submission manually - white screen and toast “Background send running, please try again shortly” which is I guess ok but how sending is ongoing as it has already failed? It is trying and trying and trying? But send fail message below form shows a precise hour when it failed and it is not changing as it would try again. When toast disappears white screen is present which is not that good, I have to choose back button o go back to collect. 
Screenshots attached: 
![screenshot_20181023-094627](https://user-images.githubusercontent.com/33342488/47357203-5ef84e00-d6c6-11e8-9510-31a0f3374370.png)
![screenshot_20181023-094907](https://user-images.githubusercontent.com/33342488/47357202-5ef84e00-d6c6-11e8-825f-45db0c9061b6.png)
In such case on master there is no even a try to autosent submissions when aggregate credentials are not input. I can see just Finalized status.

@opendatakit-bot unlabel "needs testing"

lognaturel commented:
Thank you, @kkrawczyk123. Are you satisfied that you’ve otherwise tried all the cases? 1000+ submissions, sending from an external app, things like that?

lognaturel commented:
> Notifications and action of periodic form update don’t work at all on 8.1.

Are you saying no notifications work at all or just notifications in the case of periodic form updates? Did you get to try this on Android 8.0 as well and did it work there?

> On cellular only in autosend - submission was sent when I turn wifi on not cellular and the same in the opposite situation when autosend was set to wifi only and I turn on cellular data.

Does that device also work with the correct connection type? That is, if auto-send is set to cellular only and you turn on cellular only or if auto-send is set to wifi only and you turn on wifi only?

kkrawczyk123 commented:
I tried on Androids 4.1, 4.2, 4.4, 6.0, 7.0 and 8.1.
for wifi and cellular for aggregate and google sheet.
- fill submission online and autosend
- fill submission offline, turn on connection, autosend
- fill offline, turn on connection, autosend worked on background
- killing an app
- case with airplane mode and the recovery of the connection
- tried to restart a device
- periodic form update
- I have tried to send 1000 submissions to Google sheet only once on Android 6.0 as it takes long time and I got crashed on 781 submissions probably of the same reason as https://github.com/opendatakit/collect/issues/2626 so I wasn't worried about that much.

I did not try case with an external app.

> Are you saying no notifications work at all or just notifications in the case of periodic form updates? Did you get to try this on Android 8.0 as well and did it work there?

I mean that notifications of form update. Autosend notifications worked just fine. 

> Does that device also work with the correct connection type? That is, if auto-send is set to cellular only and you turn on cellular only or if auto-send is set to wifi only and you turn on wifi only?

Yes, it worked for correct connection type as well. 

@lognaturel It to so good that you suggest me to check that https://github.com/opendatakit/collect/issues/2671#issuecomment-432253434 as I could discover another issue! When I tried those steps (not touched autosend at all) just try to send submissions to google sheet without sheet's URL - I see that toast about sending in the background and a white page as I attached above! And not only on Android 8 but on all devices!

lognaturel committed:
Remove check for ongoing auto-send from GSheets Activity

Reverts part of fa1bd1b. This never made any sense -- the check was before the API results so it would always block. Plus, what we want to do is block the upload button and that's taken care of in InstanceUploaderList.

lognaturel committed:
Close cursor after getting all forms from it

Avoid a crash now that forms are read in a loop when uploading Google Sheets instances.

Restructure code to better match InstancesDao.getInstancesFromCursor.

lognaturel commented:
Thanks for identifying these issues, @kkrawczyk123! 

I believe I fixed the blank screen on GSheets upload issue.

@grzesiek2010 took on the notification issue with 8.0+ at #2676 and it would be good to review and merge that one first. 

I've verified attempting to send 1998 submissions to an Aggregate server. All seems to be ok with that though I did fix a couple of database leak issues. Uploading a lot of submissions to GSheets hangs in a really odd way but that is the same on master so we don't have to handle it here. I suspect some of the GSheets libraries block the UI thread.

I tried basic sending from the intents tester and will be sure to tag the Ona folks in the beta so they can verify it with their real companion app.

I think there will be a conflict after #2676 is merged. I can handle that my morning and then hopefully it will be ready for merge. I think we should keep the commits on this one since a lot is moving around and it would be good to be able to track a bug to exactly where it was introduced.

lognaturel committed:
Use singleton database helper to avoid leaks

There were lots of resource leaks logged which I thought were cursors not being closed. It turns out it was actually SQLiteOpenHelper objects that were leaking. This is because d52ce0b stopped using the singleton pattern for the helper and instead started creating a new helper that was never closed with every query.

lognaturel commented:
After some more trials, I discovered more resource leaks. These took me a long time to hunt down because I thought they were cursor-related. It turns out that ~d52ce0b~  #2344 inadvertently started creating lots of unclosed database helpers. I don't think there are any risks to going back to using that structure but it would be good to do some quick checks around database access. 

In general, I think the database layer could use some deliberate thought to make sure patterns are used consistently and that we actually know resources are properly de-allocated. Hopefully this addresses at least some subset of the database issues we've been seeing in error logs.

kkrawczyk123 commented:
@lognaturel, Good news is that nearly all issues have been fixed which is great! but I am still able to reproduce an issue with Moto G Play, Android 6.0 - I am still able to autosend submission when a cellular only option is on and when wifi is on, not a cellular data and the opposite. All other devices have no problems with differentiating the type of connection.

lognaturel committed:
Merge remote-tracking branch 'upstream/master' into collect-2499

lognaturel committed:
Look in google repo before jcenter

The lint dependency wasn't being found. The recommendation is to put the google() repo before jcenter() because it contains speciality packages. https://stackoverflow.com/a/48438866/137744
